### ISSUE TITLE
How to set the distributed optimizer with a loaded Keras model

### ISSUE THREAD

**Issue by ViaFerrata** [uber/horovod#120]
Hello there, 

suppose that I've saved a horovod trained Keras model ( `model.save`) with an optimizer that has saved weights - e.g. Adam.

Now, I want to continue training by loading the model (`ks.models.load_model`).
How can I set the distributed optimizer (`hvd.DistributedOptimizer()`) in this case, since there's no reference to the loaded Adam optimizer?

The only way I can imagine now, is to modify the Keras `load_model` function in order to get a reference to the loaded optimizer.

Thanks a lot for your help!

**Comment by alsrgv** [comment]
@ViaFerrata, there are a few ways:

1. You can re-compile the model:
```python
model = keras.models.load_model('file.h5', compile=False)

# Adjust learning rate based on number of GPUs.
opt = keras.optimizers.SGD(learning_rate * hvd.size(), momentum=0.9)

# Add Horovod Distributed Optimizer.
opt = hvd.DistributedOptimizer(opt)

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=opt,
              metrics=['accuracy', 'top_k_categorical_accuracy'])
```

2. I haven't tried it, but I bet you can change optimizer after you loaded the model:
```python
model = keras.models.load_model('file.h5')

# Re-wrap model's optimizer into DistributedOptimizer:
model.optimizer = hvd.DistributedOptimizer(model.optimizer)
```

3. You should also be able to specify optimizer via custom object:
```python
model = keras.models.load_model('file.h5', custom_objects={
    'Adam': lambda **kwargs: hvd.DistributedOptimizer(keras.optimizers.Adam(**kwargs))
})
```

Which way does seem most attractive to you?

**Comment by ViaFerrata** [comment]
Thank you so much for your detailed answer!

Regarding 1), I always thought that the weights of the saved optimizer are lost if you re-compile the model. However, I could be mistaken for sure? If the recompilation doesn't change the saved config of the optimizer, this would be the simplest way of course.

Also, I didn't know the method `model.optimizer`, will try that tomorrow.

And thanks also for the custom object solution, that seems quite elegant :)

Thank you very much again for your help and of course thanks a lot for making horovod available to the public!

**Comment by alsrgv** [comment]
You're right that re-compilation resets the optimizer config - e.g. gradient momentums and other per-variable parameters.  I believe reloading the model actually has the same effect - the momentums will be reset to 0. 

Please let me know which way you end up going with, as this may be good learning for the community.

**Comment by ViaFerrata** [comment]
Are you sure about that?

If I understand the Keras FAQ (and some issues) correctly, the state of the previous optimizer is fully loaded with `load_model` (https://keras.io/getting-started/faq/):

> How can I save a Keras model?
> Saving/loading whole models (architecture + weights + optimizer state)
> You can use model.save(filepath) to save a Keras model into a single HDF5 file which will contain:
> - the architecture of the model, allowing to re-create the model
> - the weights of the model
> - the training configuration (loss, optimizer)
> - the state of the optimizer, allowing to resume training exactly where you left off.

I'll try 2) and 3) tomorrow and will report back then.
As far as I could figure it out from the horovod sources, `hvd.DistributedOptimizer` doesn't do any recompiling. So if you can access the optimizer with `model.optimizer` it should work.

**Comment by alsrgv** [comment]
AFAIK, `load_model` does reload training configuration (what loss/optimizer, learning rate, momentum rate (e.g. 0.9), etc - these are stored as `config`), but it does not reload per-variable momentums which are reset to 0 after loading the model.

Example of `config`:
```python
def get_config(self):
    config = {'lr': float(K.get_value(self.lr)),
              'momentum': float(K.get_value(self.momentum)),
              'decay': float(K.get_value(self.decay)),
              'nesterov': self.nesterov}
    base_config = super(SGD, self).get_config()
    return dict(list(base_config.items()) + list(config.items()))
```

As you can see, there are no per-variable momentums in the config.

**Comment by ViaFerrata** [comment]
Ah, now I get what you mean. Yes, there are no per-variable momenta in the config.

But I think they may be saved in the `save_model` code, after the optimizer config is saved ([Link](https://github.com/fchollet/keras/blob/master/keras/models.py#L143)):
```          
# Save optimizer weights.
symbolic_weights = getattr(model.optimizer, 'weights')
if symbolic_weights:
    optimizer_weights_group = f.create_group('optimizer_weights')
    weight_values = K.batch_get_value(symbolic_weights)
    weight_names = []
```

So for me, it looks like the `save_model` function accesses the weights attribute of the used optimizer which should also include the per-variable momenta?
However, I'm not really sure, so I've opened a Stackoverflow issue https://stackoverflow.com/questions/47716273/keras-does-save-model-really-save-all-optimizer-weights

**Comment by alsrgv** [comment]
@ViaFerrata, even if they are saved, I don't believe they're reloaded back, based on the [load_model()](https://github.com/fchollet/keras/blob/409b0005b5d3bec469009a95302e95fa8bdf2452/keras/models.py#L181) code.  Happy to be proven wrong though!

**Comment by alsrgv** [comment]
@ViaFerrata, actually, it looks like I was wrong - [this part of load_model()](https://github.com/keras-team/keras/blob/409b0005b5d3bec469009a95302e95fa8bdf2452/keras/models.py#L274) is restoring the optimizer weights.  This is very nice.  What it means is that options 2) and 3) are better than 1).

**Comment by ViaFerrata** [comment]
@alsrgv Thanks for looking into it again! Yes, I also thought that this part loads ALL (at least I hope) optimizer weights. So we should be safe.
I'll stick to using option 2) then.

Thank you very much again for your help and quick replies!

**Comment by bearnshaw** [comment]
@ViaFerrata did option 2 actually work?  I get errors like the following:
```
FailedPreconditionError (see above for traceback): Attempting to use uninitialized value SGD/iterations
```
which I think results from standard session setting lines:
```
config = tf.ConfigProto()
config.gpu_options.allow_growth = True
config.gpu_options.visible_device_list = str(hvd.local_rank())
K.set_session(tf.Session(config=config))
```

**Comment by alsrgv** [comment]
@bearnshaw, I was able to utilize (2) with this simple example, using Keras 2.1.3:

```python
import keras
import horovod.keras as hvd
import keras.backend as K
import tensorflow as tf
import numpy as np

hvd.init()

config = tf.ConfigProto()
config.gpu_options.allow_growth = True
config.gpu_options.visible_device_list = str(hvd.local_rank())
K.set_session(tf.Session(config=config))

model = keras.models.load_model('checkpoint-90.h5')
model.optimizer = hvd.DistributedOptimizer(model.optimizer)
model.fit(np.random.random([1,224,224,3]), np.random.random([1, 1000]))
```

**Comment by bearnshaw** [comment]
@alsrgv None of these options actually restores the state of the _optimizer_.  This is obvious for options 1 and 3 since new optimizers are instantiated.  Option 2 looks like it might work, however because of [this line of code](https://github.com/uber/horovod/blob/master/horovod/keras/__init__.py#L89) a new keras optimizer also gets instantiated, which for SGD [zeros out the iterations](https://github.com/keras-team/keras/blob/master/keras/optimizers.py#L164) and then calls init on the super class Optimizer which [empties the weights](https://github.com/keras-team/keras/blob/master/keras/optimizers.py#L71).  I think the correct thing to do is the following:
```
model = keras.models.load_model('file.h5')
opt = model.optimizer
model.optimizer = hvd.DistributedOptimizer(opt)
model.optimizer.iterations = opt.iterations
model.optimizer.weights = opt.weights
model.compile(optimizer=model.optimizer,
              loss='categorical_crossentropy',
              metrics=['accuracy'])
```
I think the recompile is necessary since it builds the horovod ops into the graph, but to be honest I'm not sure.

I wish that were the end of the story, but whenever I use this method to resume training two odd things happen:

1. When looking at the jumble of lines of keras training print statements, there no longer appears to be just one run being tracked like when I'm training de novo.
1. After resuming, I get a sharp decrease in training loss as if the learning rate had been dropped, even though the learning rate remains constant.

Any feedback on all of this would be helpful.

**Comment by tgaddair** [comment]
A couple initial observations:

1. Keras' optimizer saving does not save all the optimizer state.  For example, the `iterations` property used above is not actually saved along [with the other parameters](https://github.com/keras-team/keras/blob/master/keras/optimizers.py#L511).  So, currently there's no way to retrieve it on load.
2. The `fit()` method resets a number of things that will mess with any attempt to pick up where you left off.  For example, in addition to [resetting model history](https://github.com/keras-team/keras/issues/6697), every call to `fit()` will reset callbacks through the `on_train_begin()` method.  So, for example, if you made use of `ReduceLROnPlateau` to reduce the learning rate, that state will also be [reset](https://github.com/keras-team/keras/blob/2.0.4/keras/callbacks.py#L805) on every call to fit.

I did verify that `model.compile()` will not wipe out the optimizer weights if the Horovod optimizer is patched with the weights from the base optimizer as @bearnshaw showed, so his solution may be the best we can do for now.  I think it's likely that some of the issues described here contributed to the sharp decrease in training loss that was observed.  I'll look closer into the issue of different runs being tracked.

**Comment by alsrgv** [comment]
This part is very subtle, so I had to dig out some of the prior analysis.

`model.compile()` does [reset `model.train_function`](https://github.com/keras-team/keras/blob/5fcd832b5c5025b164c99f0bd46cb94d707b93d3/keras/engine/training.py#L455), and subsequent call to `fit()` would [make a new one](https://github.com/keras-team/keras/blob/409b0005b5d3bec469009a95302e95fa8bdf2452/keras/engine/training.py#L1634), which wipes optimizer state [[1]](https://github.com/keras-team/keras/blob/409b0005b5d3bec469009a95302e95fa8bdf2452/keras/engine/training.py#L977), [[2]](https://github.com/keras-team/keras/blob/409b0005b5d3bec469009a95302e95fa8bdf2452/keras/optimizers.py#L224).

Again, would be happy to be proven wrong!

**Comment by ViaFerrata** [comment]
@bearnshaw 
This might be unlikely, but are you using any layers with ```stateful``` properties (e.g. LSTM, BN)?
I've recently found out that aside from the optimizer, Keras also loses some layer information during ```model.save```. I.e., the ```layer.stateful=True``` property is not saved during the saving of the model. Thus, if you load the model again, it will be reset to the standard ```False```. 

As for the optimizer, thank you for digging so deep. Seems like there is no ideal solution right now, since Keras itself loses some optimizer information during saving such that this has to be fixed upstream.

**Comment by tgaddair** [comment]
As a test, I added the following callback to print the optimizer weights on train and epoch begin:

```
class PrintOptimizerState(keras.callbacks.Callback):
    def __init__(self, opt):
        super(PrintOptimizerState, self).__init__()
        self.opt = opt

    def on_train_begin(self, logs=None):
        print("on_train_begin: ", self.opt.get_weights())

    def on_epoch_begin(self, epoch, logs=None):
        print("on_epoch_begin: ", epoch, self.opt.get_weights())
```

As @alsrgv suggested, using method (2), the weights are zeroed out by the time we get to training, whereas method (3) retains the correct weights.  So looks like (3) is the way to go for now.

**Comment by bearnshaw** [comment]
@alsrgv @tgaddair thank you for looking into this and creating a patch so quickly!  How soon until #359 will be merged into master? And thank you for horovod, this really is a fantastic project.

@ViaFerrata not using stateful model, but thanks for the heads up!

**Comment by tgaddair** [comment]
Hey @bearnshaw, it will be sometime this week, hopefully the next day or two.  #359 is ready to go, but we also want to add a similar feature for the PyTorch API to maintain feature parity.  We'll merge the two together when that's completed.

**Comment by tgaddair** [comment]
And now version 0.13.9 has been released with these changes.  Let me know if you have any issues!

**Comment by PaulGwamanda** [comment]
For me, this solved it: 

```
import horovod.keras as hvd

model = hvd.load_model(weights)
```

Reference: https://horovod.readthedocs.io/en/stable/api.html

horovod.tensorflow.keras.load_model(filepath, custom_optimizers=None, custom_objects=None, compression=<class 'horovod.tensorflow.compression.NoneCompressor'>)[source]
Loads a saved Keras model with a Horovod DistributedOptimizer.

The DistributedOptimizer will wrap the underlying optimizer used to train the saved model, so that the optimizer state (params and weights) will be picked up for retraining.

By default, all optimizers in the module keras.optimizers will be loaded and wrapped without needing to specify any custom_optimizers or custom_objects.
